File: api/battle_sys/turn_counter.py
class TurnCount:
    def __init__(self,character_list):
        # character_list is a list of dictionary where dictionary = {'character_name': speed}
        # EX:
        #character_list = [{'phys_tb': 100},{'dan_heng'}:108]

        self.turn_list = []
        self.character_list = character_list
        self.character_speed_list = []
        self.base_AV_dict = {}
        self.cycle = 1
        self.current_cycle_AV = 0

        for char in character_list:
            self.character_speed_list.append(char.get_turn_info())

        AV_dict = {}
        for char in self.character_speed_list:
            for k,v in char.items():
                AV_dict[k] = 10000/v

        #cleaning up base AV dict values
        self.base_AV_dict = AV_dict.copy()
        #rounding up
        for k,v in self.base_AV_dict.items():
            if v.is_integer() == False:
                new_v = int(v + 1)
                self.base_AV_dict[k] = new_v

        #calculating turn order
        lowest_AV = 10000000
        lowest_char = None
        while AV_dict != {}:
            for k,v in  AV_dict.items():
                if v < lowest_AV:
                    lowest_AV = v
                    lowest_char = k

            if lowest_AV.is_integer() == False:
                lowest_AV = int(lowest_AV + 1)

            self.turn_list.append({lowest_char:int(lowest_AV)})
            AV_dict.pop(lowest_char)
            lowest_char = None
            lowest_AV = 10000000

    def set_turn_order(self):
        AV_dict = {}
        for char in self.turn_list:
            for k,v in char.items():
                AV_dict[k] = v

        self.turn_list = []
        lowest_AV = 10000000
        lowest_char = None
        while AV_dict != {}:
            for k,v in  AV_dict.items():
                if v < lowest_AV:
                    lowest_AV = float(v)
                    lowest_char = k

            if lowest_AV.is_integer() == False:
                lowest_AV = int(lowest_AV + 1)

            self.turn_list.append({lowest_char:int(lowest_AV)})

            AV_dict.pop(lowest_char)
            lowest_char = None
            lowest_AV = 10000000

    def update_cycle(self,AV):
        if self.cycle == 1:
            self.current_cycle_AV = self.current_cycle_AV + AV
            if self.current_cycle_AV >= 150:
                self.cycle += 1
                self.current_cycle_AV = self.current_cycle_AV - 150
        elif self.cycle > 1:
            self.current_cycle_AV = self.current_cycle_AV + AV
            while self.current_cycle_AV >= 100:
                self.cycle += 1
                self.current_cycle_AV = self.current_cycle_AV - 100

    def current_character(self):
        for k,v in self.turn_list[0].items():
            return k

    def next_turn(self):
        first = self.turn_list[0]
        AV_consumption = 0
        current_character = None
        for k,v in first.items():
            AV_consumption = v
            current_character = k
        self.update_cycle(AV_consumption)
        self.turn_list.pop(0)
        for char in self.turn_list:
            for k,v in char.items():
                char[k] = v - AV_consumption

        base_AV = self.base_AV_dict[current_character]
        self.turn_list.append({current_character:base_AV})
        self.set_turn_order()
        return self.turn_list


# cr_list = [
#         {'char1': 140},
#         {'char2': 103},
#         {'char3': 111},
#         {'char4': 105}
#        ]

# test = TurnCount(cr_list)
# print('Current turn order: ',test.turn_list)
# print('Current character: ',test.current_character())
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)
# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current cycle: ', test.cycle)
# print('current cycle AV: ', test.current_cycle_AV)

# print('turn taken: ',test.next_turn())
# print('turn taken: ',test.next_turn())
# print('current character taking action:',test.current_character())


# https://hsr.keqingmains.com/speed-guide/#compendium

File: ghi/src/battle_sys/turn_counter.mjs
class TurnCount{
    constructor(character_list){
        this.turn_list = []
        this.character_list = character_list
        this.character_speed_list = []
        this.base_AV_dict = {}
        this.cycle = 1
        this.current_cycle_AV = 0

        for (let char of character_list){
            this.character_speed_list.push(char.get_turn_info())
        }

        let AV_dict = {}

        for (let char of this.character_speed_list){
            for (let name in char){
                const spd = char[name]
                let av = 10000/spd
                if(Number.isInteger(av) === false){
                    av = parseInt(av) + 1
                }
                AV_dict[name] = av
            }
        }
        this.base_AV_dict = JSON.parse(JSON.stringify(AV_dict))

        while(Object.keys(AV_dict).length !== 0){
            let lowest_AV = 100000000
            let lowest_char = null
            for(let char in AV_dict){
                const spd = AV_dict[char]
                if(spd < lowest_AV){
                    lowest_AV = spd
                    lowest_char = char
                }
            }
            const char = {}
            char[lowest_char] = lowest_AV
            this.turn_list.push(char)
            delete AV_dict[lowest_char]
        }
    }
    set_turn_order(){
        let AV_dict = {}
        for(let item of this.turn_list){
            for(let char in item){
                AV_dict[char] = item[char]
            }
        }
        this.turn_list = []

        while(Object.keys(AV_dict).length !== 0){

            let lowest_AV = 10000000
            let lowest_char = null

            for(let char in AV_dict){
                const spd = AV_dict[char]
                if(spd < lowest_AV){
                    lowest_AV = spd
                    lowest_char = char
                }
            }
            const char = {}
            char[lowest_char] = lowest_AV
            this.turn_list.push(char)
            delete AV_dict[lowest_char]
        }
    }
    update_cycle(AV){
        if(this.cycle === 1){
            this.current_cycle_AV += AV
            if(this.current_cycle_AV >= 150){
                this.cycle += 1
                this.current_cycle_AV -= 150
                }
            } else if (this.cycle > 1){
                this.current_cycle_AV += AV
                while(this.current_cycle_AV >= 100){
                    this.cycle += 1
                    this.current_cycle_AV -= 100
                }
            }
        }
    current_character(){
        return Object.keys(this.turn_list[0])[0]
    }
    next_turn(){
        const first = Object.keys(this.turn_list[0])[0]
        const AV_consumption = Object.values(this.turn_list[0])[0]
        this.update_cycle(AV_consumption)
        this.turn_list.splice(0,1)

        for(let item of this.turn_list){
            for(let char in item){
                const AV = item[char]
                item[char] = AV - AV_consumption
            }
        }

        const base_AV = this.base_AV_dict[first]
        const char = {}
        char[first] = base_AV
        this.turn_list.push(char)
        this.set_turn_order()
    }
}


export default TurnCount
